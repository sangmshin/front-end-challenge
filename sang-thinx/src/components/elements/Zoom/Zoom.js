import React, { useEffect, useState } from 'react';
import { array, bool, func } from 'prop-types';
import { connect } from 'react-redux';
import { blockScroll, closeZoom } from 'store/ui/actions';
import { isZoomOpen, getZoomImages } from 'store/ui/selectors';
import { Button } from 'components/elements';
import styles from './Zoom.module.scss';
import closeButton from 'images/close-btn.svg';

export const Zoom = ({ blockScroll, closeZoom, images, isZoomOpen }) => {
  const [isOpen, setOpen] = useState(false);

  useEffect(() => {
    if (isZoomOpen && !isOpen) open();
    if (!isZoomOpen && isOpen) close();
  })

  useEffect(() => () => {
    blockScroll(false);
    isOpen && close();
  }, []);

  const keyListener = e => {
    const isEscapeButton = e.keyCode === 27;
    if (isEscapeButton) closeZoom();
  }

  const toggleKeyListener = toggle => {
    if (toggle) window.addEventListener('keyup', keyListener);
    else window.removeEventListener('keyup', keyListener);
  };
  
  const open = () => {
    toggleKeyListener(true);
    blockScroll(true);
    setOpen(true);
  };

  const close = () => {
    toggleKeyListener(false);
    blockScroll(false);
    setOpen(false);
  };

  const zoomedImages = images.map(({ src, alt }, index) => (
    <div key={index}>
      <img
        className={styles['zoom-image']}
        src={src}
        alt={alt}
        onClick={() => closeZoom()}
      />
    </div>
  ));

  if (isZoomOpen) {
    return (
      <div className={styles.root}>
        <div className={styles['close-button']}>
          <Button
            accessibleText="Close"
            ariaLabel='Close'
            onClick={() => closeZoom()}
            role='presentation'
            type='button'
          >
            <img src={closeButton} alt='close-button' />
          </Button>
        </div>
        {zoomedImages}
      </div>
    );
  }

  return null;
}

Zoom.propTypes = {
  blockScroll: func.isRequired,
  closeZoom: func.isRequired,
  images: array,
  isZoomOpen: bool,
}

Zoom.defaultProps = {
  images: [],
  isZoomOpen: false,
};

const mapStateToProps = state => ({
  images: getZoomImages(state),
  isZoomOpen: isZoomOpen(state)
});


const mapDispatchToProps = {
  closeZoom,
  blockScroll
};

export default connect(mapStateToProps, mapDispatchToProps)(Zoom);