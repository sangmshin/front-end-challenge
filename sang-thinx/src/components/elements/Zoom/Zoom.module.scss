@import 'variables';

.root {
  z-index: $image-zoom;
  position: fixed;
  top: 0px;
  bottom: 0px;
  left: 0px;
  right: 0px;
  overflow: scroll;

  .close-button {
    position: fixed;
    top: 20px;
    right: 30px;
    z-index: 1;
    cursor: pointer;
  }

  .zoom-image {
    width: 100%;
    cursor: pointer;
    float: left;
  }
}