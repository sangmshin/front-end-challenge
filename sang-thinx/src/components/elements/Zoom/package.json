{
  "name": "Zoom",
  "main": "./Zoom.js"
}
