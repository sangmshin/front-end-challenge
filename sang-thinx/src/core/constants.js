export const COLOR_HEXCODE_MAP = {
  beige: '#f2d9a5',
  black: '#000',
}