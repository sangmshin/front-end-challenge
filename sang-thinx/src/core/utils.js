const SMALL = 768; /* Small screen / tablet */
const MEDIUM = 1024; /* Medium screen / desktop */
const LARGE = 1200; /* Large screen / wide desktop */
let BREAKPOINT;

/**
 * Finds the current matched media query breakpoint
 *
 * @returns {String} matched media query breakpoint [large, medium, small, xsmall]
 */
export function getCurrentBreakPoint() {
  if (!BREAKPOINT) {
    BREAKPOINT = {
      large: window.matchMedia(`(min-width: ${LARGE}px)`),
      medium: window.matchMedia(`(min-width: ${MEDIUM}px)`),
      small: window.matchMedia(`(min-width: ${SMALL}px)`),
      xsmall: window.matchMedia(`(max-width: ${SMALL - 1}px)`)
    };
  }

  return Object.keys(BREAKPOINT).find(breakpoint => BREAKPOINT[breakpoint].matches);
}

/**
 * Determines if the data is array and has at least one or more items
 * @return {boolean}
 */
export const isArrayWithLength = arr => Array.isArray(arr) && arr.length > 0;

/**
 * A function with no operation. (for testing)
 */
export const noop = () => null;
