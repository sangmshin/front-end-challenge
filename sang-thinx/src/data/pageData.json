
{
  "template": "product-detail",
  "hero": {
    "images": [
      {
        "src": "https://thinx-assets.s3.amazonaws.com/highwaist_black_front_01_1024x1024.jpg", "alt": "Hi-Waist front 1"
      },
      {
        "src": "https://thinx-assets.s3.amazonaws.com/highwaist_black_front_02_1024x1024.jpg", "alt": "Hi-Waist front 2"
      },
      {
        "src": "https://thinx-assets.s3.amazonaws.com/highwaist_black_front_03_1024x1024.jpg", "alt": "Hi-Waist front 3"
      },
      {
        "src": "https://thinx-assets.s3.amazonaws.com/highwaist_black_front_04_1024x1024.jpg", "alt": "Hi-Waist front 4"
      },
      {
        "src": "https://thinx-assets.s3.amazonaws.com/highwaist_black_side_1024x1024.jpg", "alt": "Hi-Waist side"
      },
      {
        "src": "https://thinx-assets.s3.amazonaws.com/highwaist_black_back_1024x1024.jpg", "alt": "Hi-Waist back"
      }
    ],
    "product_title": "Hi - Waist",
    "product_droplets": "Heavy days. Holds up to 2 tampon's worth",
    "description": "The phrase \"treat yo'self\" reaches new belly-button brushing heights with the Hi-Waist, and it's here just in time for cuffing season.",
    "price": "35.00",
    "available_colors": [
      "black",
      "beige"
    ],
    "available_sizes": [
      "xsmall",
      "small",
      "medium",
      "large",
      "xlarge",
      "xxlarge"
    ],
    "size_chart": true,
    "savings": {
      "label": "build a cycle set and save up to 20%",
      "url": ""
    }
  },
  "best_sellers": [
    {
      "product_title": "Hiphugger",
      "product_droplets": "heavy days",
      "image": "https://thinx-assets.s3.amazonaws.com/best_seller_1.jpg"
    },
    {
      "product_title": "Cheeky",
      "product_droplets": "light days",
      "image": "https://thinx-assets.s3.amazonaws.com/best_seller_2.jpg"
    },
    {
      "product_title": "Thong",
      "product_droplets": "lightest days",
      "image": "https://thinx-assets.s3.amazonaws.com/best_seller_3.jpg"
    },
    {
      "product_title": "Sport",
      "product_droplets": "medium days",
      "image": "https://thinx-assets.s3.amazonaws.com/best_seller_4.jpg"
    },
    {
      "product_title": "Boyshort",
      "product_droplets": "light days",
      "image": "https://thinx-assets.s3.amazonaws.com/best_seller_5.jpg"
    }
  ]
}
