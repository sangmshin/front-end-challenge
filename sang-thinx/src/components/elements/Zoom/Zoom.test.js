import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Zoom } from './Zoom';
import { noop } from 'core/utils';
import { hero } from 'data/pageData.json';

const props = {
  blockScroll: noop,
  closeZoom: noop,
  images: hero.images,
  isZoomOpen: true
}

describe('Component :: Zoom', () => {
  describe('#render', () => {
    const component = shallow(
      <Zoom {...props} />
    );
    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('#isZoomOpen prop', () => {
    const component = shallow(
      <Zoom {...props} isZoomOpen={false} />
    );
    it('should not render anything when isZoomOpen prop from redux is set to false', () => {
      expect(component.getElement()).toBe(null);
    });
  });

  describe('#click in the option', () => {
    const spy = jest.fn();
    const component = shallow(
      <Zoom {...props} closeZoom={spy} />
    );
    it('should executes the onSelect', () => {
      component.find('ForwardRef').simulate('click');
      expect(spy).toHaveBeenCalled();
    });
  });
});
