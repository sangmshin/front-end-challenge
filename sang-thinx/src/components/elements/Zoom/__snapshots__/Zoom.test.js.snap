// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: Zoom #render should match snapshot 1`] = `
<div
  className="root"
>
  <div
    className="close-button"
  >
    <ForwardRef
      accessibleText="Close"
      ariaLabel="Close"
      className={null}
      disabled={false}
      onClick={[Function]}
      role="presentation"
      style={null}
      type="button"
    >
      <img
        alt="close-button"
        src="close-btn.svg"
      />
    </ForwardRef>
  </div>
  <div
    key="0"
  >
    <img
      alt="Hi-Waist front 1"
      className="zoom-image"
      onClick={[Function]}
      src="https://thinx-assets.s3.amazonaws.com/highwaist_black_front_01_1024x1024.jpg"
    />
  </div>
  <div
    key="1"
  >
    <img
      alt="Hi-Waist front 2"
      className="zoom-image"
      onClick={[Function]}
      src="https://thinx-assets.s3.amazonaws.com/highwaist_black_front_02_1024x1024.jpg"
    />
  </div>
  <div
    key="2"
  >
    <img
      alt="Hi-Waist front 3"
      className="zoom-image"
      onClick={[Function]}
      src="https://thinx-assets.s3.amazonaws.com/highwaist_black_front_03_1024x1024.jpg"
    />
  </div>
  <div
    key="3"
  >
    <img
      alt="Hi-Waist front 4"
      className="zoom-image"
      onClick={[Function]}
      src="https://thinx-assets.s3.amazonaws.com/highwaist_black_front_04_1024x1024.jpg"
    />
  </div>
  <div
    key="4"
  >
    <img
      alt="Hi-Waist side"
      className="zoom-image"
      onClick={[Function]}
      src="https://thinx-assets.s3.amazonaws.com/highwaist_black_side_1024x1024.jpg"
    />
  </div>
  <div
    key="5"
  >
    <img
      alt="Hi-Waist back"
      className="zoom-image"
      onClick={[Function]}
      src="https://thinx-assets.s3.amazonaws.com/highwaist_black_back_1024x1024.jpg"
    />
  </div>
</div>
`;
